### DISCUSSION TITLE
How can I set the custom Filter for AutoCompleteBox?

### DISCUSSION THREAD

**Discussion by kyuranger** [AvaloniaUI/Avalonia#11392] [Q&A] [answered]
What I want to do is show all the items while the text is empty, and filter other text as FilterMode Contains does.

It seems currently no FilterMode can do this so I have to set FilterMode to custom for creating a custom Filter.

There is no tutorial about this, could anyone please help me?

Thank you.

**Comment by sccombp** [comment]
Hi kyuranger, I believe this is a bug in the AutoCompleteBox where it will not display its dropdown if no characters have been entered.

There is a behaviour you can use that works around this issue which you can find [here](https://github.com/AvaloniaUI/Avalonia/issues/8903#issuecomment-1331594149) which will open the dropdown when you press the down arrow or click F4.

Using this behaviour will allow you to use the standard FilterMode 'Contains' without having to create a custom filter.

something like this (please excuse my bad sample code):
```XML
<UserControl xmlns="https://github.com/avaloniaui"
             xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
             xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
             xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
             xmlns:bh="using:Werk.Behaviours"
             xmlns:i="clr-namespace:Avalonia.Xaml.Interactivity;assembly=Avalonia.Xaml.Interactivity"
             mc:Ignorable="d" d:DesignWidth="800" d:DesignHeight="450"
             x:Class="test">

  <AutoCompleteBox
         Items="{Binding YourItemList}"
         IsTextCompletionEnabled="True"
         FilterMode="Contains">
    <i:Interaction.Behaviors>
      <bh:AutoCompleteZeroMinimumPrefixLengthDropdownBehaviour/>
    </i:Interaction.Behaviors>
  </AutoCompleteBox>

</UserControl>
```

**Reply by kyuranger** [reply]
> Hi kyuranger, I believe this is a bug in the AutoCompleteBox where it will not display its dropdown if no characters have been entered.
> 
> There is a behaviour you can use that works around this issue which you can find [here](https://github.com/AvaloniaUI/Avalonia/issues/8903#issuecomment-1331594149) which will open the dropdown when you press the down arrow or click F4.
> 
> Using this behaviour will allow you to use the standard FilterMode 'Contains' without having to create a custom filter.
> 
> something like this (please excuse my bad sample code):
> 
> ```
> <UserControl xmlns="https://github.com/avaloniaui"
>              xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
>              xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
>              xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
>              xmlns:bh="using:Werk.Behaviours"
>              xmlns:i="clr-namespace:Avalonia.Xaml.Interactivity;assembly=Avalonia.Xaml.Interactivity"
>              mc:Ignorable="d" d:DesignWidth="800" d:DesignHeight="450"
>              x:Class="test">
> 
>   <AutoCompleteBox
>          Items="{Binding YourItemList}"
>          IsTextCompletionEnabled="True"
>          FilterMode="Contains">
>     <i:Interaction.Behaviors>
>       <bh:AutoCompleteZeroMinimumPrefixLengthDropdownBehaviour/>
>     </i:Interaction.Behaviors>
>   </AutoCompleteBox>
> 
> </UserControl>
> ```
I added your behaviours with your sample but it doesn't work any. When I press F4 it throws error.

**Reply by sccombp** [reply]
my apologies, that code was just to show you how to attach the behaviour to the AutoCompleteBox. I have attached a sample project

[TestAutoCompleteBehaviour.zip](https://github.com/AvaloniaUI/Avalonia/files/11494077/TestAutoCompleteBehaviour.zip)

**Reply by kyuranger** [reply]
> my apologies, that code was just to show you how to attach the behaviour to the AutoCompleteBox. I have attached a sample project
> 
> [TestAutoCompleteBehaviour.zip](https://github.com/AvaloniaUI/Avalonia/files/11494077/TestAutoCompleteBehaviour.zip)

It works now. But I met a strange situation. I want to show the drop-down list when I press the textbox everytime. For this I add a
 ```
protected override void OnAttached()
        {
            if (AssociatedObject is not null)
            {
                AssociatedObject.KeyUp += OnKeyUp;
                AssociatedObject.DropDownOpening += DropDownOpening;
                AssociatedObject.PointerPressed += AssociatedObject_PointerPressed;                
            }

            base.OnAttached();
        }

        private void AssociatedObject_PointerPressed(object? sender, Avalonia.Input.PointerPressedEventArgs e)
        {
            if (string.IsNullOrEmpty(AssociatedObject?.Text))
            {
                ShowDropdown();
            }
        }
```
in protected override void OnAttached().

However, it doesn't work any. When I change it from PointerPressed to GotFocus, it only works while first time click it.

What's wrong with it.

**Reply by sccombp** [reply]
Change it to PointerReleased instead.

**Reply by kyuranger** [reply]
> Change it to PointerReleased instead.

Well, it works. Could you please tell me difference between PointerReleased and PointerPressed?

**Reply by sccombp** [reply]
PointerReleased is when you click and release (no longer interacting with the control), this allows the control to display its dropdown.

PointerPressed is when you click but have not yet released, this prevents the control from displaying the dropdown as you are keeping focus on the textbox (and the dropdown automatically closes if it does not have focus)

**Reply by kyuranger** [reply]
> PointerReleased is when you click and release (no longer interacting with the control), this allows the control to display its dropdown.
> 
> PointerPressed is when you click but have not yet released, this prevents the control from displaying the dropdown as you are keeping focus on the textbox (and the dropdown automatically closes if it does not have focus)

Thank you so much.

**Reply by kyuranger** [reply]
> PointerReleased is when you click and release (no longer interacting with the control), this allows the control to display its dropdown.
> 
> PointerPressed is when you click but have not yet released, this prevents the control from displaying the dropdown as you are keeping focus on the textbox (and the dropdown automatically closes if it does not have focus)

I met a new problem. Your code works in stable version 0.10.19. But in the newest version 11.0.0 Preview8 it can't works.

It throws the error in following code:
 ```
private void ShowDropdown()
        {
            if (AssociatedObject is not null && !AssociatedObject.IsDropDownOpen)
            {
                typeof(AutoCompleteBox).GetMethod("PopulateDropDown", System.Reflection.BindingFlags.NonPublic | System.Reflection.BindingFlags.Instance)?.Invoke(AssociatedObject, new object[] { AssociatedObject, EventArgs.Empty });
                typeof(AutoCompleteBox).GetMethod("OpeningDropDown", System.Reflection.BindingFlags.NonPublic | System.Reflection.BindingFlags.Instance)?.Invoke(AssociatedObject, new object[] { false });

                //opening the dropdown does not set this automatically.
                //We *must* set the field and not the property as we need to avoid the changed event being raised (which prevents the dropdown opening).
                var iddo = typeof(AutoCompleteBox).GetField("_isDropDownOpen", System.Reflection.BindingFlags.NonPublic | System.Reflection.BindingFlags.Instance);
                if ((bool)iddo?.GetValue(AssociatedObject) == false)
                    iddo?.SetValue(AssociatedObject, true);
            }
        }
```
The error is :

> System.NullReferenceException
>   HResult=0x80004003
>   Message=Object reference not set to an instance of an object.
>   Source=AvaloniaApplication2
>   StackTrace:
>    at AvaloniaApplication2.Behaviours.AutoCompleteZeroMinimumPrefixLengthDropdownBehaviour.ShowDropdown() in C:\Users\Admin\Desktop\AvaloniaApplication2\Behaviours\AutoCompleteZeroMinimumPrefixLengthDropdownBehaviour.cs:line 82
> 
>   This exception was originally thrown at this call stack:
>     AvaloniaApplication2.Behaviours.AutoCompleteZeroMinimumPrefixLengthDropdownBehaviour.ShowDropdown() in AutoCompleteZeroMinimumPrefixLengthDropdownBehaviour.cs

It seems because the code:
`var iddo = typeof(AutoCompleteBox).GetField("_isDropDownOpen", System.Reflection.BindingFlags.NonPublic | System.Reflection.BindingFlags.Instance);`

gets a null result.

**Reply by sccombp** [reply]
I'm afraid it looks like the changes made [here](https://github.com/AvaloniaUI/Avalonia/commit/10d41eaeee6bf2ae9822905c45746a9ef0730643) have removed the ability to set the isDropDownOpen value without triggering a property changed event. 

I don't really have the time to look further into this right now, hopefully a better solution will present itself before 11 goes final.

**Reply by kyuranger** [reply]
> I'm afraid it looks like the changes made [here](https://github.com/AvaloniaUI/Avalonia/commit/10d41eaeee6bf2ae9822905c45746a9ef0730643) have removed the ability to set the isDropDownOpen value without triggering a property changed event.
> 
> I don't really have the time to look further into this right now, hopefully a better solution will present itself before 11 goes final.

All right, thank you all the same.

**Reply by sccombp** [reply, marked as answer]
good news! I managed to get it working for v11
you'll need to change the ShowDropdown procedure to 

```
        private void ShowDropdown()
        {
            if (AssociatedObject is not null && !AssociatedObject.IsDropDownOpen)
            {
                typeof(AutoCompleteBox).GetMethod("PopulateDropDown", System.Reflection.BindingFlags.NonPublic | System.Reflection.BindingFlags.Instance)?.Invoke(AssociatedObject, new object[] { AssociatedObject, EventArgs.Empty });
                typeof(AutoCompleteBox).GetMethod("OpeningDropDown", System.Reflection.BindingFlags.NonPublic | System.Reflection.BindingFlags.Instance)?.Invoke(AssociatedObject, new object[] { false });

                if (!AssociatedObject.IsDropDownOpen)
                {
                    //We *must* set the field and not the property as we need to avoid the changed event being raised (which prevents the dropdown opening).
                    var ipc = typeof(AutoCompleteBox).GetField("_ignorePropertyChange", System.Reflection.BindingFlags.NonPublic | System.Reflection.BindingFlags.Instance);
                    if ((bool)ipc?.GetValue(AssociatedObject) == false)
                        ipc?.SetValue(AssociatedObject, true);

                    AssociatedObject.SetCurrentValue<bool>(AutoCompleteBox.IsDropDownOpenProperty, true);
                }
            }
        }
```

**Reply by kyuranger** [reply]
> good news! I managed to get it working for v11 you'll need to change the ShowDropdown procedure to
> 
> ```
>         private void ShowDropdown()
>         {
>             if (AssociatedObject is not null && !AssociatedObject.IsDropDownOpen)
>             {
>                 typeof(AutoCompleteBox).GetMethod("PopulateDropDown", System.Reflection.BindingFlags.NonPublic | System.Reflection.BindingFlags.Instance)?.Invoke(AssociatedObject, new object[] { AssociatedObject, EventArgs.Empty });
>                 typeof(AutoCompleteBox).GetMethod("OpeningDropDown", System.Reflection.BindingFlags.NonPublic | System.Reflection.BindingFlags.Instance)?.Invoke(AssociatedObject, new object[] { false });
> 
>                 if (!AssociatedObject.IsDropDownOpen)
>                 {
>                     //We *must* set the field and not the property as we need to avoid the changed event being raised (which prevents the dropdown opening).
>                     var ipc = typeof(AutoCompleteBox).GetField("_ignorePropertyChange", System.Reflection.BindingFlags.NonPublic | System.Reflection.BindingFlags.Instance);
>                     if ((bool)ipc?.GetValue(AssociatedObject) == false)
>                         ipc?.SetValue(AssociatedObject, true);
> 
>                     AssociatedObject.SetCurrentValue<bool>(AutoCompleteBox.IsDropDownOpenProperty, true);
>                 }
>             }
>         }
> ```

I just tested it. It works now! Thank you.

**Reply by pyking-wang** [reply]
I want to disable the filter and display all the dropdown options when I click the dropdown button. Can I do this？

**Reply by sccombp** [reply]
I think you meant to add a reply to [this](https://github.com/AvaloniaUI/Avalonia/discussions/11889) thread instead.

Try changing the CreateDropdownButton code in the behaviour to something like this (all we're doing is now clearing the autocompletebox before triggering the dropdown)

```csharp
        private void CreateDropdownButton()
        {
            if (AssociatedObject != null)
            {
                var prop = AssociatedObject.GetType().GetProperty("TextBox", System.Reflection.BindingFlags.Instance | System.Reflection.BindingFlags.NonPublic);
                var tb = (TextBox?)prop?.GetValue(AssociatedObject);
                
                var btn = new Button()
                {
                    Content = "🢗",
                    Margin = new(3),
                    ClickMode = ClickMode.Press
                };
                btn.Click += (s, e) => 
                {
                    AssociatedObject.Text = string.Empty;
                    ShowDropdown(); 
                };

                tb.InnerRightContent = btn;
            }
        }
`

**Reply by pyking-wang** [reply]
Yes, it's cool，thanks.

**Reply by pyking-wang** [reply]
There is a small issue. After `AssociatedObject. Text=string. Empty,` the text on the autocompletebox's textbox does not appear empty, but it is actually empty, and an item must be selected to not be empty, which is a bit strange
![image](https://github.com/AvaloniaUI/Avalonia/assets/76192669/2c0c86b5-144f-4cb4-a4c1-3aee3549f2a0)

**Reply by timunie** [reply]
I believe there was a PR to improve it. #12057 Might be on nightly only, so check nightly builds if possible.

**Reply by pyking-wang** [reply]
I have updated nightly version，but the problem has not been solved=-=
![image](https://github.com/AvaloniaUI/Avalonia/assets/76192669/debc8f5e-d58e-4165-ac88-ac281cb6a579)
A not very good way to avoid this problem
